### PR TITLE
Websocket server that streams events from client.

### PR DESCRIPTION
**Pull request by jgraef**
## Pull request checklist

- [X] All tests pass. Demo project builds and runs.
- [X] I have resolved any merge conflicts.

## What's in this pull request?

Added a Websocket server that streams events to connected clients.

## ToDo:

 * Authentication
 * IP white-listing
 * Accepting TX over websocket stream

## Notes:

Somehow changes to the `ValidatorPool` sneaked in. I need to work on the `ConnectionPool` and it's interface to the Validator code anyway. So this will be fixed later.

### CODE REVIEW

File: lib/src/extras/metrics_server.rs [hunk 1]
```diff
@@ -8,14 +8,18 @@ use crate::config::consts::default_bind;
 
 
 pub fn initialize_metrics_server(client: &Client, config: MetricsServerConfig) -> Result<MetricsServer, Error> {
+    let ip = config.bind_to.unwrap_or_else(default_bind);
+    info!("Initializing metrics server: {}:{}", ip, config.port);
+
     let (username, password) = if let Some(credentials) = config.credentials {
         (Some(credentials.username), Some(credentials.password))
     } else {
+        warn!("No password set for RPC server!");
```
Comment (paberr) [inline lib/src/extras/metrics_server.rs]: Should be metrics server here

File: network-primitives/src/networks.rs [hunk 2]
```diff
@@ -152,7 +152,7 @@ lazy_static! {
             network_id: NetworkId::DevAlbatross,
             name: "dev-albatross",
             seed_peers: vec![
-                //create_seed_peer_addr_ws("albatross.nimiq.dev", 8444, "5af4c3f30998573e8d3476cd0e0543bf7adba576ef321342e41c2bccc246c377"),
+                create_seed_peer_addr_ws("albatross.nimiq.dev", 8444, "5af4c3f30998573e8d3476cd0e0543bf7adba576ef321342e41c2bccc246c377"),
```
Comment (paberr) [inline network-primitives/src/networks.rs]: Do we want this?

File: lib/src/extras/ws_rpc_server.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+use crate::config::consts::default_bind;
+
+pub fn initialize_ws_rcp_server(client: &Client, config: WsRpcServerConfig) -> Result<WsRpcServer, Error> {
+    let ip = config.bind_to.unwrap_or_else(default_bind);
+
+    info!("Initializing websocket RPC server: {}:{}", ip, config.port);
+
+    let server = WsRpcServer::new(ip, config.port)?;
+    server.register_blockchain(client.consensus());
+    if let Some(validator) = client.validator() {
```
Comment (paberr) [inline lib/src/extras/ws_rpc_server.rs]: This needs to be behind a feature flag.

**Review by paberr — CHANGES_REQUESTED** [review]
Few changes required.

File: network-primitives/src/networks.rs (hunk 2 above)
Comment (jgraef) [reply network-primitives/src/networks.rs]: Oops. No

File: lib/src/extras/ws_rpc_server.rs (hunk 3 above)
Comment (jgraef) [reply lib/src/extras/ws_rpc_server.rs]: No, because the whole module is behind a feature flag.

Comment (jgraef) [reply lib/src/extras/ws_rpc_server.rs]: The validator part must be behind a feature gate. Fixed it.

**Review by paberr — APPROVED** [review]

**Pull request merged**